### DISCUSSION TITLE
iptables vs. nft

### DISCUSSION THREAD

**Discussion by feckert** [strongswan/strongswan#1019] [General]
Openwrt uses in the upcoming stable release openwrt-22.03 and in the master branch, nft as the default firewall back-end.
So that strongswan puts the iptables/ip6tables into the nft framework and not in the legacy iptables backend, `iptables-nft/ip6tables-nft ` should be used. Openwrt installs as default the `iptable-nft/ip6tables-nft` if old syntax is needed.

Strongswan does not support native nft rules, So we must use iptablen-nft/ip6tables-nft for the `strongswan-plugin-updown`.
I have changed this, so that the dependency is fulfilled.
The iptables-legacy/ip6tables-legacy packages is not installed with this [change](https://github.com/openwrt/packages/commit/fdeaa02ccfe69929fc137253e835fa61c62ec585) as default.

My problem now is that the plugins `strongswan-plugin-connmark` and `strongswan-mod-forecast` use the legacy firewall via internal libip4tc lib.
https://github.com/openwrt/packages/blob/master/net/strongswan/Makefile#L673
https://github.com/openwrt/packages/blob/master/net/strongswan/Makefile#L690

My question is: How can these plugins be changed so that they use new nft firewall backend and not the old iptables firewall backend.

**Comment by tobiasbrunner** [comment]
> My question is: How can these plugins be changed so that they use new nft firewall backend and not the old iptables firewall backend.

I guess they eventually have to be rewritten/adapted/extended to configure nft via Netlink (if available, otherwise fallback to libiptc if that's available/linked).

**Reply by feckert** [reply]
> > My question is: How can these plugins be changed so that they use new nft firewall backend and not the old iptables firewall backend.
> 
> I guess they eventually have to be rewritten/adapted/extended to configure nft via Netlink (if available, otherwise fallback to libiptc if that's available/linked).

@tobiasbrunner Thank you for your answer :+1:

The plugin ` strongswan-plugin-updown` works, because in the script the iptables(ipv4) or ip6tables(ipv6) command is called directly.
A symlink decides which iptables backend is used.

- iptables point either to `iptables-nft` or `iptables-legacy`(ipv4)
- ip6tables point either to `ip6tables-nft` or `ip6tables-legacy`(ipv6)

But if we use libip4tc/libip6tc for the plugins `strongswan-plugin-connmark` or `strongswan-plugin-forecast` then the rules are created always in the old firewall backend, right? As far as I know, either **nft** or **iptables** should be used and not both of the at the same time.

This means for openwrt as long as the modules do not use the new nft, I cannot use this  strongswan plugins for openwrt.
Unless I use the old firewall backend or we port this to the new nft backend.

@tobiasbrunner Do you know any other modules where we have the same problem?

**Reply by tobiasbrunner** [reply]
> The plugin strongswan-plugin-updown works, because in the script the iptables(ipv4) or ip6tables(ipv6) command is called directly.

If `iptables-nft` is able to convert all the rules correctly that should work.

> As far as I know, either nft or iptables should be used and not both of the at the same time.

I don't know exactly what the technical implications are of mixing the two (as long as the kernel is built with support for xtables and libiptc  is available the plugins might continue to work). But even if that's not really an issue, it could be confusing to users if they don't see the rules installed by the plugins (`iptables/iptables-nft` will not show them, they'd have to call `iptables-legacy` if it's still available).

> Do you know any other modules where we have the same problem?

What do you mean?

**Reply by feckert** [reply]
> > The plugin strongswan-plugin-updown works, because in the script the iptables(ipv4) or ip6tables(ipv6) command is called directly.
> 
> If `iptables-nft` is able to convert all the rules correctly that should work.
> 

yes that works. But I'm not sure yet how the sequence for working the rules will be.
OpenWrt uses its own firewall [table](https://git.openwrt.org/?p=project/firewall4.git;a=summary) and the iptables-nft rules are added in a separate other table in nft.
Now I'm still trying to figure out which firewall table to go through first.

> > As far as I know, either nft or iptables should be used and not both of the at the same time.
> 
> I don't know exactly what the technical implications are of mixing the two (as long as the kernel is built with support for xtables and libiptc is available the plugins might continue to work).

Yes, that is not a problem. The module should compile. But whether it will work as intended when nft is used is uncertain!

> > Do you know any other modules where we have the same problem?
> 
> What do you mean?

Are there any other modules in the strongswan that use iptables and need to be switched to nft so that the strongswan works with both iptabes and nft?

**Reply by tobiasbrunner** [reply]
> Are there any other modules in the strongswan that use iptables and need to be switched to nft so that the strongswan works with both iptabes and nft?

I see. No, the default updown script and the _connmark_ and _forecast_ plugins are the only ones.
